### PR TITLE
Add `MergeDeep` type

### PR DESCRIPTION
**Pull request by skarab42**
@sindresorhus Sorry for the extra work, I'm a bit embarrassed :( I need to review the basics of git and especially conflict resolution. Fixes #443

### CODE REVIEW

**Comment by sindresorhus** [comment]
There are some unresolved comments in the previous pull request:

- https://github.com/sindresorhus/type-fest/pull/443#discussion_r957058900
- https://github.com/sindresorhus/type-fest/pull/443#discussion_r957059558

**Comment by sindresorhus** [comment]
Don't forget to add `Fixes` to the pull request description.

**Comment by skarab42** [comment]
As mentioned here https://github.com/sindresorhus/type-fest/pull/443#discussion_r959225107 I noticed that all the options prefixed with `merge-or-` are the options that allow you to go deep, so maybe they should be called `deep-replace`, `deep-spread`, etc...

**Comment by sindresorhus** [comment]
> I have to do the same thing in the deep version.

Let me know when this is done.

**Comment by skarab42** [comment]
> Let me know when this is done.

@sindresorhus  Done, but I think since the test run on `type-fest@2.19.0` it fail because it does not incorporate the latest changes on which this type depends aka `OmitIndexSignature` and `PickIndexSignature`.

(P.S: sorry I didn't mean to edit your comment)

**Comment by skarab42** [comment]
@sindresorhus `Merge` is broken :( I do a fix!

**Comment by skarab42** [comment]
@sindresorhus  Well, by introducing more and more fixes and options into this type, it becomes too complicated and will quickly become unmaintainable in the long run. I'm thinking of going back to a simpler and more optimised version that doesn't try to cover every edge cases.

**Comment by sindresorhus** [comment]
I agree

**Comment by skarab42** [comment]
@sindresorhus  In blood and sweat I finally got something I like. Unfortunately I can't make it any simpler without sacrificing features that I consider essential:

- Preservation and merging of index signatures.
- Preservation and normalisation of optional properties.
- Merging of variable length tuples.
- Recursion into arrays/tuples.

**Comment by sindresorhus** [comment]
This is looking much better. Thanks for working so hard on it.

What's the reasoning for choosing `spread` as default `arrayMergeMode`? Isn't `replace` more expected behavior as that's how it behaves in native JS when you merge (using object.assign or object spread).

I struggle to see when I would use the `union` `arrayMergeMode`. Do you have any real-world use-cases in mind?

**Comment by skarab42** [comment]
> Isn't `replace` more expected behavior as that's how it behaves in native JS when you merge (using object.assign or object spread).

This is indeed the case when you merge two objects, but when you merge two arrays it's the opposite. So I wonder if we shouldn't export a `MergeDeepArray` type with `spread` by default and `MergeDeepObject` with `replace` by default.

> I struggle to see when I would use the `union` `arrayMergeMode`.

```ts
const initialState = {data: ['a', 'b', 'c']};

// Do the merge
let nextState = {...initialState, data: [1, 2, 3]};

// Error when we try to revert back to inital state, data is typed as number[]
nextState = initialState;
```

**Comment by sindresorhus** [comment]
> This is indeed the case when you merge two objects, but when you merge two arrays it's the opposite. So I wonder if we shouldn't export a MergeDeepArray type with spread by default and MergeDeepObject with replace by default.

Yeah, the important distinction is whether the array is top-level or deep in an object. When it's top-level, it makes to spread, when it's in an object, it makes sense to replace by default.

> In fact I'm not really convinced of the usefulness of union the above example is not realistic as we're going to type the state correctly from the beginning.

Maybe drop it for now and see if there's any requests for it?

**Comment by skarab42** [comment]
@sindresorhus I hope this time it's all right.

**Comment by skarab42** [comment]
> I also don't see when you would want `arrayMergeMode: 'replace'` for top-level arrays. 

Probably when used in another type.

> So maybe it's easier to make it so that top-level arrays are always spread and document that `arrayMergeMode` only applies to nested arrays?

@sindresorhus  Yep right. I added the `spreadTopLevelArrays` option with a default value of `true` which is the easiest solution I found without having to create a new branch and duplicate some code.

**Comment by sindresorhus** [comment]
> Probably when used in another type.

Not sure I understand. Do you have a use-case in mind?

> I added the spreadTopLevelArrays option with a default value of true which is the easiest solution I found without having to create a new branch and duplicate some code.

My question is, do we even need an option for it?

---

Sorry this is taking long. I'm just trying hard to keep the type as simple as possible and ensure we don't add stuff people don't actually need.

**Comment by skarab42** [comment]
> My question is, do we even need an option for it?

We don't need to make it public, the vast majority of users won't need it (and for the moment I can't find any specific use cases). I can make it static but that won't simplify the type itself, only it signature.

**Comment by skarab42** [comment]
@sindresorhus Alternatively, we could also mark the option as `@internal`?

**Comment by sindresorhus** [comment]
> I can make it static but that won't simplify the type itself, only it signature.

👍

**Comment by skarab42** [comment]
@sindresorhus Done ;)

File: source/merge-deep.d.ts [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+			: never
+>;
+
+/**
+Merge two objects or two arrays/tuples recursively into a new type.
+
+- Properties that only exist in one object are copied into the new object.
+- Properties that exist in both objects are merged if possible or replaced by the one of the source if not.
+- Top-level arrays and tuples are always spread.
+- By default, inner arrays and tuples are replaced. See {@link MergeDeepOptions.arrayMergeMode arrayMergeMode} option to change this behaviour.
```
Comment (sindresorhus) [inline source/merge-deep.d.ts:364]: ^ Here it says `replaced`. But here it says `spread`: https://github.com/sindresorhus/type-fest/pull/452/files#diff-5ce63d84bd0497ef6dd72ef89147e7bf80f464a1185b1a34139c6c1cb6c37ee5R307

**Comment by sindresorhus** [comment]
Awesome work! I'm very happy with what we ended up with ❤️

**Pull request merged**